clustercontroller-reindexer/src/main/java/ai/vespa/reindexing/Reindexer.java

```diff
@@ -0,0 +1,216 @@
+        switch (code) {
+            default:
+                log.log(WARNING, "Unexpected visitor result '" + control.getResult().getCode() + "'");
+            case FAILURE: // Intentional fallthrough — this is an error.
+                log.log(WARNING, "Visiting failed: " + control.getResult().getMessage());
+                return status.failed(clock.instant(), control.getResult().getMessage());
+            case ABORTED:
+                log.log(FINE, () -> "Halting reindexing of " + type + " due to shutdown — will continue later");
+                return status.halted();
+            case SUCCESS:
+                log.log(INFO, "Completed reindexing of " + type + " after " + Duration.between(status.startedAt(), clock.instant()));
+                return status.successful(clock.instant());
+        }
+    }
+
+    private void visit(DocumentType type, ProgressToken progress, VisitorControlHandler control) {
+        VisitorParameters parameters = createParameters(type, progress);
+        parameters.setControlHandler(control);
+        VisitorSession session;
+        try {
+            session = access.createVisitorSession(parameters);
+        }
+        catch (ParseException e) {
+            throw new IllegalStateException(e);
+        }
+
+        // Wait until done, or interrupted, in which case we abort the visit but don't wait for it to complete.
+        try {
+            control.waitUntilDone();
+        }
+        catch (InterruptedException e) {
+            control.abort();
+            Thread.currentThread().interrupt();
+        }
+        session.destroy(); // If thread is interrupted, this will not wait, but will retain the interrupted flag.
+    }
+
+    VisitorParameters createParameters(DocumentType type, ProgressToken progress) {
+        VisitorParameters parameters = new VisitorParameters(type.getName());
+        parameters.setRemoteDataHandler(cluster.name());
+        parameters.setResumeToken(progress);
+        parameters.setFieldSet(type.getName() + ":[document]");
+        parameters.setPriority(DocumentProtocol.Priority.LOW_1);
+        parameters.setRoute(cluster.route());
+        parameters.setBucketSpace(cluster.bucketOf(type));
+        // parameters.setVisitorLibrary("ReindexVisitor");
+        return parameters;
+    }
+
+
+    static class Cluster {
+
+        private final String name;
+        private final String configId;
+        private final Map<DocumentType, String> documentBuckets;
+
+        Cluster(String name, String configId, Map<DocumentType, String> documentBuckets) {
+            this.name = requireNonNull(name);
+            this.configId = requireNonNull(configId);
+            this.documentBuckets = Map.copyOf(documentBuckets);
+        }
+
+        String name() {
+            return name;
+        }
+
+        String route() {
+            return "[Storage:cluster=" + name + ";clusterconfigid=" + configId + "]";
+        }
+
+        String bucketOf(DocumentType documentType) {
```

Comment:
I'd prefer if this were called `bucketSpaceOf`, as "bucket space" to "bucket" is a bit like "table" is to "tablet" in e.g. BigTable, i.e. on different levels of abstraction. Consider also renaming other variables to match.

Reply:
Yes yes yes. Obviously ;)